test/e2e/alertmanager_test.go

```diff
@@ -161,3 +170,220 @@ func TestAlertmanagerTrustedCA(t *testing.T) {
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		err = framework.Poll(5*time.Second, 5*time.Minute, func() error {
+			secrets, err := f.KubeClient.CoreV1().Secrets(testNs).List(metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			for _, secret := range secrets.Items {
+				_, dockerToken := secret.Annotations["openshift.io/create-dockercfg-secrets"]
+				token := strings.Contains(secret.Name, fmt.Sprintf("%s-token-", sa))
+
+				// we have to skip the token secret that contains the openshift.io/create-dockercfg-secrets annotation
+				// as this is the token to talk to the internal registry.
+				if !dockerToken && token {
+					clients[sa] = silenceClient{
+						t:         t,
+						host:      host,
+						token:     string(secret.Data["token"]),
+						namespace: testNs,
+					}
+					return nil
+				}
+			}
+			return errors.Errorf("couldn't find token secret for service account %q", sa)
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Create a silence.
+	now := time.Now()
+	sil := []byte(fmt.Sprintf(
+		`{"matchers":[{"name":"foo","value":"bar","isRegex":false}],"startsAt":"%s","endsAt":"%s","createdBy":"somebody","comment":"some comment"}`,
+		now.Format(time.RFC3339),
+		now.Add(time.Hour).Format(time.RFC3339),
+	))
+
+	for _, sa := range []string{"viewer", "anonymous"} {
+		t.Logf("Creating silence as %q (denied)", sa)
+		_ = clients[sa].do("POST", "/api/v2/silences", sil, http.StatusForbidden)
+	}
+
+	t.Log("Creating silence as 'editor' (allowed)")
+	b := clients["editor"].do("POST", "/api/v2/silences", sil, http.StatusOK)
+
+	// Save silence ID for deletion.
+	parsed, err := gabs.ParseJSON(b)
+	if err != nil {
+		t.Fatalf("%v (data: %q)", err, string(b))
+	}
+	silID, ok := parsed.Path("silenceID").Data().(string)
+	if !ok {
+		t.Fatalf("couldn't get silenceID from response %q", string(b))
+	}
+
+	// List silences.
+	t.Logf("Listing silences as 'anonymous' (denied)")
+	clients["anonymous"].do("GET", "/api/v2/silences", nil, http.StatusForbidden)
+
+	for _, sa := range []string{"viewer", "editor"} {
+		t.Logf("Listing silences as %q (allowed)", sa)
+		b = clients[sa].do("GET", "/api/v2/silences", nil, http.StatusOK)
+
+		parsed, err = gabs.ParseJSON(b)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		count, err := parsed.ArrayCount()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if count != 1 {
+			t.Fatalf("expecting 1 silence, got %d (%q)", count, string(b))
+		}
+	}
+
+	// Delete the silence.
+	for _, sa := range []string{"viewer", "anonymous"} {
+		t.Logf("Deleting silence as %q (denied)", sa)
+		_ = clients[sa].do("DELETE", fmt.Sprintf("/api/v2/silence/%s", silID), nil, http.StatusForbidden)
+	}
+
+	t.Log("Deleting silence as 'editor' (allowed)")
+	_ = clients["editor"].do("DELETE", fmt.Sprintf("/api/v2/silence/%s", silID), sil, http.StatusOK)
+}
+
+// The Alertmanager API should be protected by the OAuth proxy.
+func TestAlertmanagerOAuthProxy(t *testing.T) {
```

Comment:
Can you explain how this is testing alertmanger oauth proxy, we just do a query against alertmanager.

Reply:
We're querying Alertmanager via its OpenShift route which *should* be mapped to the OAuth proxy service. This is copied from https://github.com/openshift/cluster-monitoring-operator/blob/e597c67470412d5094d162f415c8b0d7cd3fe7f9/test/e2e/user_workload_monitoring_test.go#L376-L403

I should probably add an e2e test checking that a request without sufficient permissions is denied.